File: src/main/scala/com/vc/medium/L529.scala
package com.vc.medium

object L529 {
  def updateBoard(board: Array[Array[Char]], click: Array[Int]): Array[Array[Char]] = {
    val dirs = Array(
      (0, 1),
      (1, 1),
      (1, 0),
      (1, -1),
      (0, -1),
      (-1, -1),
      (-1, 0),
      (-1, 1)
    )
    val x = click(0)
    val y = click(1)
    if(board(x)(y) == 'M') {
      board(x)(y) = 'X'
    }
    else {
      import scala.collection.mutable
      val st = new mutable.Stack[(Int, Int)]()
      st.push((x, y))
      while(st.nonEmpty) {
        val e = st.pop
        var mine = 0
        val stInner = new mutable.Stack[(Int, Int)]()
        dirs.foreach(dir => {
          val xNew = e._1 + dir._1
          val yNew = e._2 + dir._2
          if(xNew >= 0 && yNew >= 0 && xNew < board.length && yNew < board(0).length) {
            if(board(xNew)(yNew) == 'M') mine += 1
            else if(board(xNew)(yNew) == 'E') stInner.push((xNew, yNew))
          }
        })
        if(mine > 0) {
          board(e._1)(e._2) = (mine + '0').toChar
        }
        else {
          board(e._1)(e._2) = 'B'
          while(stInner.nonEmpty) st.push(stInner.pop)
        }
      }
    }
    board
  }
}

File: src/main/java/com/vc/medium/L529.java
package com.vc.medium;

import java.util.*;

class L529 {

    private int[][] dirs = {{0, 1},{1, 1},{1, 0},{1, -1},{0, -1},{-1, -1},{-1, 0},{-1, 1}};
    private char[][] board;
    private Stack<int[]> intermediate = new Stack<int[]>();

    public char[][] updateBoard(char[][] board, int[] click) {
        this.board = board;
        Stack<int[]> st = new Stack<>();
        st.push(click);

        while(!st.isEmpty()) {
            int[] e = st.pop();
            int row = e[0], col = e[1];

            if(board[row][col] == 'M') board[row][col] = 'X';
            else {
                intermediate.clear();
                int count = countAdjacentMines(row, col);
                if(count > 0) board[row][col] = (char)(count + '0');
                else {
                    board[row][col] = 'B';
                    st.addAll(intermediate);
                }
            }
        }
        return board;
    }

    private int countAdjacentMines(int row, int col) {
        int count = 0;
        for(int[] dir: dirs) {
            int rowNew = dir[0] + row;
            int colNew = dir[1] + col;
            if(rowNew >= 0 && rowNew < board.length && colNew >= 0 && colNew < board[0].length) {
                if(board[rowNew][colNew] == 'X' || board[rowNew][colNew] == 'M') count++;
                else if(board[rowNew][colNew] == 'E') intermediate.add(new int[]{rowNew, colNew});
            }
        }
        return count;
    }

}
